### PR TITLE
M1 - Impl json serialization types for ExternalTransition

### LINKED ISSUE #228: [M1] Implement json serialization for enum types under External transition
**Issue summary**

Part of #214

**Other information and links**

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
This PR adds a derive macro `AutoFrom` that removes ~1100 lines of boilerplate code under `base/src/serialization_type_conversions`

example: 

```rust
#[derive(Debug, Clone, PartialEq, derive_more::From, derive_more::Into)]
struct I64(pub i64);

#[derive(Debug, Clone, PartialEq, mina_serialization_types_macros::AutoFrom)]
#[auto_from(Bar)]
enum Foo {
    V1,
    V2(i64),
    V3(i64, Option<i64>, Vec<i64>, Box<i64>),
    V4 {
        f1: i64,
    },
    V5 {
        f1: i64,
        f2: Option<i64>,
        f3: Vec<i64>,
        f4: Box<i64>,
    },
}

#[derive(Debug, Clone, PartialEq)]
enum Bar {
    V1,
    V2(I64),
    V3(I64, Option<I64>, Vec<I64>, Box<I64>),
    V4 {
        f1: I64,
    },
    V5 {
        f1: I64,
        f2: Option<I64>,
        f3: Vec<I64>,
        f4: Box<I64>,
    },
}

type BarV1 = versioned::Versioned<Bar, 1>;
let foo = Foo::V5 {
    f1: 8,
    f2: Some(9),
    f3: vec![10, 11, 12],
    f4: Box::new(13),
};

let bar: Bar = foo.clone().into();
let bar_v1: BarV1 = foo.clone().into();
let foo_from_bar: Foo = bar.into();
let foo_from_bar_v1: Foo = bar_v1.into();
assert_eq!(foo, foo_from_bar);
assert_eq!(foo, foo_from_bar_v1);
```

Changes introduced in this pull request:
- Adds a derive macro `AutoFrom` that helps generate boilerplate code of type conversions when the 2 types have identical field names and their fields are convertible between each other.
- Implement `ConsensusStateJson` which is json serialization type for `ConsensusState`
- Implement `ExternalTransitionJson` which is json serialization type for `ExternalTransition`
- json binprot parity test for entire block
- Removes boilerplate code for type conversions
- Removes `Base58Encodable` and `Base64Encodable`, implement string conversions (`TryFrom<String>` and `TryFrom<$ty> for String>`) with json types.

**Reference issue to close (if applicable)**

Closes #228 #214 

**Other information and links**

### CODE REVIEW

File: protocol/serialization-types-macros/tests/roundtrip.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(test)]
+mod tests {
+    use mina_serialization_types_macros::*;
+
+    #[derive(Debug, Clone, PartialEq, derive_more::From, derive_more::Into)]
+    struct I64(pub i64);
```
Comment (hanabi1224) [inline protocol/serialization-types-macros/tests/roundtrip.rs:9]: This macro crate is the major part for review

File: consensus/tests/consensus_state.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(test)]
+mod tests {
+    use mina_rs_base::{types::ConsensusState, JsonSerializationType};
+    use test_fixtures::*;
+
+    #[test]
+    fn consensus_state_json_serde_tests() -> anyhow::Result<()> {
```
Comment (hanabi1224) [inline consensus/tests/consensus_state.rs:10]: @creativcoder  json serde tests for `ConsensusState` is here

File: base/src/account/permissions.rs [hunk 3]
```diff
@@ -39,16 +43,16 @@ pub struct Permissions {
...
-            edit_state: AuthRequired::Signature,
-            send: AuthRequired::Signature,
-            receive: AuthRequired::None,
-            set_delegate: AuthRequired::Signature,
-            set_permissions: AuthRequired::Signature,
-            set_verification_key: AuthRequired::Signature,
-        }
-    }
-}
+// impl Default for Permissions {
```
Comment (creativcoder) [inline base/src/account/permissions.rs:46]: nit: Remove this?

File: base/src/consensus_state.rs [hunk 4]
```diff
@@ -56,7 +47,9 @@ impl AsRef<[u8]> for VrfOutputTruncated {
 /// approach where the future stake distribution snapshot is prepared by the current consensus epoch.
 ///
 /// Samasika prepares the past for the future! This future state is stored in the next_epoch_data field.
-#[derive(Clone, Debug, PartialEq, SmartDefault)]
+#[derive(Clone, Debug, PartialEq, SmartDefault, AutoFrom)]
+#[auto_from(mina_serialization_types::consensus_state::ConsensusState)]
+// #[auto_from(mina_serialization_types::consensus_state::ConsensusStateJson)]
```
Comment (creativcoder) [inline base/src/consensus_state.rs:52]: nit: uncomment this? or remove?

File: base/src/protocol_state.rs [hunk 5]
```diff
@@ -46,8 +47,9 @@ impl Hashable for ProtocolConstants {
     }
 }
 
-#[derive(Clone, Default, Debug, PartialEq, Serialize)]
-#[serde(into = "mina_serialization_types::v1::ProtocolStateV1")]
+#[derive(Clone, Default, Debug, PartialEq, AutoFrom)]
+#[auto_from(mina_serialization_types::protocol_state::ProtocolState)]
+// #[serde(into = "mina_serialization_types::v1::ProtocolStateV1")]
```
Comment (creativcoder) [inline base/src/protocol_state.rs:52]: nit: uncomment this? or remove?

File: consensus/tests/common.rs [hunk 6]
```diff
@@ -242,17 +242,29 @@ mod tests {
         let mut chain_a = ProtocolStateChain::default();
         let mut consensus_state = ConsensusState::default();
         consensus_state.blockchain_length = Length(117896);
-        consensus_state.last_vrf_output =
-            VrfOutputTruncated::from("ZYmm1mr8vJ6F-1pOKgmF1yAF41Z9onq1Je5PoKSBAwA=");
+        // BUGBUG: This should be base64 decoded
```
Comment (creativcoder) [inline consensus/tests/common.rs:245]: Yes, fixing it as part of recent changes in https://github.com/ChainSafe/mina-rs/pull/172

Comment (creativcoder) [reply consensus/tests/common.rs:245]: blake2b decoded

Comment (creativcoder) [reply consensus/tests/common.rs:245]: It's a blake2b hash btw: https://github.com/MinaProtocol/mina/blob/compatible/docs/specs/consensus/README.md#516-hashlastvrf

File: consensus/tests/density.rs [hunk 7]
```diff
@@ -37,8 +38,10 @@ mod tests {
             slots_per_epoch: Length(7140),
         };
 
+        println!("{}", chain_a.clone().try_into_json().unwrap());
```
Comment (creativcoder) [inline consensus/tests/density.rs:41]: replace this with assert is_ok maybe?

Comment (hanabi1224) [reply consensus/tests/density.rs]: This is covered by other tests, will just remove this line

File: base/src/account/permissions.rs (hunk 3 above)
Comment (hanabi1224) [reply base/src/account/permissions.rs]: fixed

File: base/src/protocol_state.rs (hunk 5 above)
Comment (hanabi1224) [reply base/src/protocol_state.rs]: fixed

File: base/src/consensus_state.rs (hunk 4 above)
Comment (hanabi1224) [reply base/src/consensus_state.rs]: fixed

File: protocol/serialization-types/src/type_annotations.rs [hunk 8]
```diff
@@ -1,19 +1,49 @@
...
 
 /// This trait annotates a given type its corresponding json serialization type,
 /// and provide utility functions to easily convert between them
-pub trait JsonSerializationType: Sized {
+pub trait JsonSerializationType<'de>: Sized {
     /// The corresponding serialization type for json format
     /// Self type can be used here to indicate no special convertion is needed
     /// TODO: Add default value when the feature lands on stable rust
-    type T: From<Self> + Into<Self>;
+    type T: From<Self> + Into<Self> + Serialize + Deserialize<'de>;
```
Comment (creativcoder) [inline protocol/serialization-types/src/type_annotations.rs:36]: Doesn't `From<Self>` imply `Into<Self>` ? do we need `Into<Self>`?

Comment (hanabi1224) [reply protocol/serialization-types/src/type_annotations.rs:36]: I believe From<Self> imply Into<Other>

Comment (creativcoder) [reply protocol/serialization-types/src/type_annotations.rs:36]: Yeah, just checked, From<T> can also be a lossy conversion. Need to have both

File: protocol/serialization-types-macros/src/auto_from.rs [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+                        }
+                        "Option" => {
+                            option_field_idents.push(ident);
+                            continue 'outer;
+                        }
+                        "Box" => {
+                            boxed_field_idents.push(ident);
+                            continue 'outer;
+                        }
+                        _ => {}
```
Comment (creativcoder) [inline protocol/serialization-types-macros/src/auto_from.rs:55]: Would be good to log the ident here, for debugging purposes

File: protocol/serialization-types-macros/src/auto_from.rs [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+                            pos_token_stream
+                                .push(quote! {item.#pos.map(::std::convert::Into::into)});
+                            continue 'outer;
+                        }
+                        "Box" => {
+                            pos_token_stream
+                                .push(quote! {::std::boxed::Box::new((*item.#pos).into())});
+                            continue 'outer;
+                        }
+                        _ => {}
```
Comment (creativcoder) [inline protocol/serialization-types-macros/src/auto_from.rs:133]: Would be good to log the ident here, for debugging purposes

File: protocol/serialization-types-macros/src/auto_from.rs [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+
+        impl<const V1: u16, const V2: u16, const V3: u16, const V4: u16> ::std::convert::From<#ident> for ::versioned::Versioned<::versioned::Versioned<::versioned::Versioned<::versioned::Versioned<#target_type, V1>, V2>, V3>, V4> {
+            #[inline]
+            fn from(t: #ident) -> Self {
+                let t: #target_type = t.into();
+                t.into()
+            }
+        }
+
+        impl<const V1: u16, const V2: u16, const V3: u16, const V4: u16> ::std::convert::From<::versioned::Versioned<::versioned::Versioned<::versioned::Versioned<::versioned::Versioned<#target_type, V1>, V2>, V3>, V4>> for #ident {
```
Comment (creativcoder) [inline protocol/serialization-types-macros/src/auto_from.rs:360]: can we type alias this somewhere?

File: protocol/serialization-types-macros/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+            _ => {}
+        },
+        syn::Data::Enum(DataEnum { variants, .. }) => {
+            if let Some(ts) =
+                auto_from::auto_from_for_enum(&ident, target_types.as_slice(), variants)
+            {
+                return ts;
+            }
+        }
+        _ => {}
```
Comment (creativcoder) [inline protocol/serialization-types-macros/src/lib.rs:119]: would be good to log for debugging purposes

File: protocol/serialization-types/src/common.rs [hunk 13]
```diff
@@ -16,15 +17,77 @@ pub type Hash2V1 = Versioned<HashV1, 1>;
...
+        ))
+    }
+}
+
+/// u64 wrapper (json)
+/// Note that integers are represented as string in mina json
+#[derive(Clone, Debug, PartialEq, derive_more::From)]
+pub struct U64Json(pub u64);
+
+impl Serialize for U64Json {
```
Comment (creativcoder) [inline protocol/serialization-types/src/common.rs:61]: I wonder if this can be shortened with: https://serde.rs/variant-attrs.html#serialize_with

**Review by creativcoder — APPROVED** [review]
Left a few comments, looks great, a lot of boiler plate removed with the macro. Thanks for this!

**Review by sudo-shashank — APPROVED** [review]

File: protocol/serialization-types-macros/src/lib.rs (hunk 12 above)
Comment (hanabi1224) [reply protocol/serialization-types-macros/src/lib.rs:119]: Changed to `unimplemented!()`, it makes more sense to just panic here

File: protocol/serialization-types/src/common.rs (hunk 13 above)
Comment (hanabi1224) [reply protocol/serialization-types/src/common.rs:61]: Unfortunately, this only saves 1 line or 2 but it requires another mod and it's not straightforward for ppl who read the code to understand what is going on unless they are very familiar with serde attributes

File: protocol/serialization-types-macros/src/auto_from.rs (hunk 9 above)
Comment (hanabi1224) [reply protocol/serialization-types-macros/src/auto_from.rs:55]: This default variant here is the most common case that falls to the default handler, adding `println!` here will flood build output

File: protocol/serialization-types-macros/src/auto_from.rs (hunk 10 above)
Comment (hanabi1224) [reply protocol/serialization-types-macros/src/auto_from.rs:133]: ditto

**Pull request merged**